OpenRA.Game/Traits/World/ScreenMap.cs

```diff
@@ -88,6 +91,41 @@ public void Remove(Actor a)
 			partitionedActors.Remove(a);
 		}
 
+		public void Add(IEffect effect, WPos position, Size size)
+		{
+			var screenPos = worldRenderer.ScreenPxPosition(position);
+			var screenBounds = new Rectangle(screenPos.X - size.Width / 2, screenPos.Y - size.Height / 2, size.Width, size.Height);
+			partitionedEffects.Add(effect, screenBounds);
+		}
+
+		public void Add(IEffect effect, WPos position, Sprite sprite)
+		{
+			var size = new Size((int)sprite.Size.X, (int)sprite.Size.Y);
+			Add(effect, position, size);
+		}
+
+		public void Update(IEffect effect, WPos position, Size size)
+		{
+			if (!partitionedEffects.Contains(effect))
+				return;
+
+			var screenPos = worldRenderer.ScreenPxPosition(position);
+			var screenBounds = new Rectangle(screenPos.X - size.Width / 2, screenPos.Y - size.Height / 2, size.Width, size.Height);
+			partitionedEffects.Update(effect, screenBounds);
+		}
+
+		public void Update(IEffect effect, WPos position, Sprite sprite)
+		{
+			var size = new Size((int)sprite.Size.X, (int)sprite.Size.Y);
+			Update(effect, position, size);
+		}
+
+		public void Remove(IEffect e)
+		{
+			if (e != null && partitionedEffects.Contains(e))
```

Comment:
This `Contains` check is unnecessary work: `Remove` *should* be safe to call when the object isn't in the collection.

Reply:
This isn't strictly true right now because `partitionedEffects.Remove` will call `MutateBins` which is more work than necessary.  You should be able to avoid that by calling `itemBounds.Remove` first and then mutating only if something was removed.

Reply:
> then mutating only if something was removed

How do I check if something was removed?

Reply:
The `Remove` method returns true if it successfully removed an item.  But on second thought this request is probably too risky unless @RoosterDragon can first confirm that my assumption about `MutateBins` only needing to be called if an element was actually removed.

Reply:
SpatiallyPartitioned.Remove as written is indeed unsafe to call with items not already in the collection.

You could easily change it to 
```c#
public bool Remove(T item)
{
	if (!itemBounds.TryGetValue(item, out Rectangle bounds))
		return false;

	MutateBins(item, bounds, removeItem);
	itemBounds.Remove(item);
	return true;
}
```

This would have the same semantics as other collections, where you can remove things that weren't necessarily there to being with (and the return result allows you to check)

Reply:
The above code doesn't compile as-is, the `if` line triggers `) expected`, `; expected` and some more.

Reply:
Sorry, declare `Rectangle bounds;` on the line prior.